---
title: 如何使用 try-catch 处理异常 - C# 编程指南
description: 了解如何使用 try-catch 块处理异常。 查看代码示例和其他可用资源。
ms.date: 07/20/2015
helpviewer_keywords:
- exception handling [C#], try/catch blocks
- exceptions [C#], try/catch blocks
- try/catch blocks [C#]
ms.assetid: ca8e3773-980e-4767-8633-7408540e9818
ms.openlocfilehash: 357aebe042bc5b6e761b3c1bad258021441de22c
ms.sourcegitcommit: 6f58a5f75ceeb936f8ee5b786e9adb81a9a3bee9
ms.translationtype: HT
ms.contentlocale: zh-CN
ms.lasthandoff: 07/28/2020
ms.locfileid: "87301991"
---
# <a name="how-to-handle-an-exception-using-trycatch-c-programming-guide"></a><span data-ttu-id="45a4f-104">如何使用 try/catch 处理异常（C# 编程指南）</span><span class="sxs-lookup"><span data-stu-id="45a4f-104">How to handle an exception using try/catch (C# Programming Guide)</span></span>
<span data-ttu-id="45a4f-105">[try-catch](../../language-reference/keywords/try-catch.md) 块的用途是捕获并处理工作代码产生的异常。</span><span class="sxs-lookup"><span data-stu-id="45a4f-105">The purpose of a [try-catch](../../language-reference/keywords/try-catch.md) block is to catch and handle an exception generated by working code.</span></span> <span data-ttu-id="45a4f-106">某些异常可以在 `catch` 块中进行处理，问题得以解决并不再出现异常；但是，大多数情况下你唯一可做的是确保引发的异常是合理异常。</span><span class="sxs-lookup"><span data-stu-id="45a4f-106">Some exceptions can be handled in a `catch` block and the problem solved without the exception being re-thrown; however, more often the only thing that you can do is make sure that the appropriate exception is thrown.</span></span>  
  
## <a name="example"></a><span data-ttu-id="45a4f-107">示例</span><span class="sxs-lookup"><span data-stu-id="45a4f-107">Example</span></span>  
 <span data-ttu-id="45a4f-108">在此示例中，<xref:System.IndexOutOfRangeException> 不是最合理的异常：<xref:System.ArgumentOutOfRangeException> 对于此方法来说更有意义，因为此错误是由调用方传递的 `index` 参数引起的。</span><span class="sxs-lookup"><span data-stu-id="45a4f-108">In this example, <xref:System.IndexOutOfRangeException> is not the most appropriate exception: <xref:System.ArgumentOutOfRangeException> makes more sense for the method because the error is caused by the `index` argument passed in by the caller.</span></span>  
  
 [!code-csharp[csProgGuideExceptions#5](~/samples/snippets/csharp/VS_Snippets_VBCSharp/csProgGuideExceptions/CS/Exceptions.cs#5)]  
  
## <a name="comments"></a><span data-ttu-id="45a4f-109">注释</span><span class="sxs-lookup"><span data-stu-id="45a4f-109">Comments</span></span>  
 <span data-ttu-id="45a4f-110">引发异常的代码包含在 `try` 块中。</span><span class="sxs-lookup"><span data-stu-id="45a4f-110">The code that causes an exception is enclosed in the `try` block.</span></span> <span data-ttu-id="45a4f-111">在此块后面紧挨着添加 `catch` 语句以处理 `IndexOutOfRangeException` 异常（如果发生此异常）。</span><span class="sxs-lookup"><span data-stu-id="45a4f-111">A `catch` statement is added immediately after to handle `IndexOutOfRangeException`, if it occurs.</span></span> <span data-ttu-id="45a4f-112">`catch` 块处理 `IndexOutOfRangeException` 异常并改为引发更合理的 `ArgumentOutOfRangeException` 异常。</span><span class="sxs-lookup"><span data-stu-id="45a4f-112">The `catch` block handles the `IndexOutOfRangeException` and throws the more appropriate `ArgumentOutOfRangeException` exception instead.</span></span> <span data-ttu-id="45a4f-113">为了向调用方提供尽可能多的信息，请考虑将原始异常指定为新异常的 <xref:System.Exception.InnerException%2A>。</span><span class="sxs-lookup"><span data-stu-id="45a4f-113">In order to provide the caller with as much information as possible, consider specifying the original exception as the <xref:System.Exception.InnerException%2A> of the new exception.</span></span> <span data-ttu-id="45a4f-114">因为 <xref:System.Exception.InnerException%2A> 属性为 [read-only](../../properties.md#read-only)，所以必须在新异常的构造函数中指定此属性。</span><span class="sxs-lookup"><span data-stu-id="45a4f-114">Because the <xref:System.Exception.InnerException%2A> property is [read-only](../../properties.md#read-only), you must assign it in the constructor of the new exception.</span></span>  
  
## <a name="see-also"></a><span data-ttu-id="45a4f-115">另请参阅</span><span class="sxs-lookup"><span data-stu-id="45a4f-115">See also</span></span>

- [<span data-ttu-id="45a4f-116">C# 编程指南</span><span class="sxs-lookup"><span data-stu-id="45a4f-116">C# Programming Guide</span></span>](../index.md)
- [<span data-ttu-id="45a4f-117">异常和异常处理</span><span class="sxs-lookup"><span data-stu-id="45a4f-117">Exceptions and Exception Handling</span></span>](./index.md)
- [<span data-ttu-id="45a4f-118">异常处理</span><span class="sxs-lookup"><span data-stu-id="45a4f-118">Exception Handling</span></span>](./exception-handling.md)
